src/core/sensor/qgsiodevicesensor.h

```diff
@@ -0,0 +1,259 @@
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
+#ifndef QGSIODEVICESENSOR_H
+#define QGSIODEVICESENSOR_H
+
+#include "qgis_core.h"
+#include "qgis_sip.h"
+#include "qgsabstractsensor.h"
+
+#include <QBuffer>
+#include <QDomElement>
+#include <QSerialPort>
+#include <QTcpSocket>
+#include <QUdpSocket>
+
+/**
+ * \ingroup core
+ * \class QgsIoDeviceSensor
+ * \brief An abstract class QIODevice-based sensor classes
+ * \since QGIS 3.32
+ */
+class CORE_EXPORT QgsIODeviceSensor : public QgsAbstractSensor
+{
+
+    Q_OBJECT
+
+  public:
+
+    explicit QgsIODeviceSensor( QObject *parent = nullptr ) : QgsAbstractSensor( parent ) {}
+    ~QgsIODeviceSensor() override;
+
+    /**
+     * Initiates the \a device.
+     */
+    void initiateIODevice( QIODevice *device );
+
+  protected slots:
+
+    void parseData() override;
+
+  private:
+
+    QIODevice *mIODevice = nullptr;
+
+};
+
+/**
+ * \ingroup core
+ * \class QgsTcpSocketSensor
+ * \brief A TCP socket sensor class
+ * \since QGIS 3.32
+ */
+class CORE_EXPORT QgsTcpSocketSensor : public QgsIODeviceSensor
+{
+
+    Q_OBJECT
+
+  public:
+
+    /**
+     * Constructor for a TCP socket sensor, bound to the specified \a parent.
+     */
+    explicit QgsTcpSocketSensor( QObject *parent = nullptr );
+    ~QgsTcpSocketSensor() override = default;
+
+    /**
+     * Returns a new TCP socket sensor.
+     *
+     * The caller takes responsibility for deleting the returned object.
+     */
+    static QgsTcpSocketSensor *create( QObject *parent ) SIP_FACTORY;
+
+    int type() const override;
+
+    /**
+     * Returns the host name the socket connects to.
+     */
+    QString hostName() const;
+
+    /**
+     * Sets the host name the socket connects to.
+     * \param hostName the host name string (a domain name or an IP address)
+     */
+    void setHostName( const QString &hostName );
+
+    /**
+     * Returns the port the socket connects to.
+     */
+    int port() const;
+
+    /**
+     * Sets the \a port the socket connects to.
+     */
+    void setPort( int port );
+
+    bool writePropertiesToElement( QDomElement &element, QDomDocument &document ) const override;
+    bool readPropertiesFromElement( const QDomElement &element, const QDomDocument &document ) override;
+
+  protected:
+
+    void handleConnect() override;
+    void handleDisconnect() override;
+
+  private slots:
+
+    virtual void socketStateChanged( const QAbstractSocket::SocketState socketState );
+
+  private:
+
+    QTcpSocket *mTcpSocket = nullptr;
```

Comment:
unique_ptr ?

Reply:
Not needed, ownership passed onto QIODeviceSensor, which handles deletion (I've moved that parent QIODevice pointer to unique_ptr).

Reply:
What about if someone calls "setIODevice" on a QgsTcpSocketSensor object -- that'll delete mIODevice but mTcpSocket will remain as a pointer to the deleted object and cause crashes.

Maybe setIODevice should be virtual and overridden in QgsTcpSocketSensor to handle this (and ensure that only QTcpSocket devices can be used when calling QgsTcpSocketSensor::setIODevice )